tests/ui/needless_move.rs

```diff
@@ -0,0 +1,254 @@
+#![warn(clippy::needless_move)]
+
+#[derive(Copy, Clone)]
+struct Copy;
+
+struct NonCopy;
+
+struct Composite {
+    copy: Copy,
+    non_copy: NonCopy,
+}
+
+impl Composite {
+    fn new() -> Self {
+        Self {
+            copy: Copy,
+            non_copy: NonCopy,
+        }
+    }
+}
+
+fn with_owned<T>(_: T) {}
+fn with_ref<T>(_: &T) {}
+fn with_ref_mut<T>(_: &mut T) {}
+
+fn main() {
+    // doesn't trigger on non-move closures or async blocks
+    let a = NonCopy;
+    let b = Copy;
+    let closure = || {
+        with_owned(a);
+        with_owned(b);
+    };
+
+    let a = NonCopy;
+    let b = Copy;
+    let fut = async {
+        with_owned(a);
+        with_owned(b);
+    };
+
+    // triggers on move closures and async blocks which do not capture anything
+    let closure = move || {};
+    let fut = async move {};
+
+    // owned + NonCopy
+    let a = NonCopy;
+    let closure = move || {
+        with_owned(a);
+    };
+
+    // owned + Copy
+    let a = Copy;
+    let closure = move || {
```

Comment:
I don't think we should lint here. Removing the `move` in this case makes the closure no longer `'static`:
```rs
fn assert_static<T: 'static>(_: T) {}

let a = Copy;
assert_static(move || { with_owned(a); }); // removing the `move` causes a compiler error
```

Reply:
Hmm that's an interesting case. I thought `Copy` types always get moved into the closure, even without `move`, but it seems I was wrong.

Reply:
Should be fixed now + I made all the test cases `assert_static` in 06d74077679adbc7ac3c9ad7040a618aa8d2b2b0